### DISCUSSION TITLE
Alternative unit and range quantity

### DISCUSSION THREAD

**Discussion by ziofat** [cooklang/spec#84] [Ideas]
I'd like to include alternative quantity of my ingredients with imperal unit. 

For example I have a stock recipe requires 1 kilogram of water, but I also want to show it in a more common way. Instead of `Add @water{1%kg}` I want to have `Add @water{1%kg, 1.25%cup}`.

Also, supporting range quantity of ingredients may help my recipe more readable and understandable. The recipe requires 700 gram chicken wings and I want to show how many chicken wings it would be normally. Instead of `Add @chicken wings{700%g}` I use `Add @chicken wings{700%g, 6-8}` to indicate 700 grams may require 6 to 8 chicken wings.

The syntax may be inappropriate but I am glad to see this feature implemented.

**Comment by shicks** [comment]
This would also be useful to address a pet peeve of mine when reading recipes: calling for "1 can" of something.  I'm always careful to annotate "1 15-oz can of ..." in my recipes, but there doesn't currently seem to be a good way to write this.  `@tomato sauce{15%oz}` reads poorly because it seems like you need to measure it out, `@can of tomato sauce{15%oz}` makes that a little clearer, but "can" really wants to be part of the unit, not the ingredient.  And `@tomato sauce{15%oz can}` is also weird, and you might end up with a shopping list calling for a "44 oz can" or something strange like that.  Instead, I'd like to write `@tomato sauce{15%oz, 1 medium can}` (or maybe the other order) and see the recipe rendered as "1 medium can tomato sauce (15 oz)" or such like.

**Reply by ziofat** [reply]
Another method is to define a unit converter for each ingredient. There can be a ingredient metadata file in project.

**Comment by adsilcott** [comment]
This is very important, and my biggest issue with cooklang so far. Many good baking recipes are written with both volume measurements and weight measurements, i.e. 2 cups (260 grams) flour, since it's much more accurate to use a kitchen scale and measure by weight, but you want to include the volume measurements for people who aren't used to baking that way and might not have a kitchen scale. 

Also, if the recipe only has volume measurements, then I'd like to be able to add weight measurements without overwriting the original recipe.

**Reply by tino** [reply]
Shouldn't the renderer just have a switch for weight or volume units and do the conversion for you?

**Reply by adsilcott** [reply]
That sounds cool, but I'm not sure how it would work. There would have to be a conversion for every type of ingredient, since one cup of flour doesn't weigh the same as one cup of grains/seeds/etc.

**Comment by jphastings** [comment]
We already have `|` pipes as synonyms for items in the shopping list spec, I think it'd be great to use for the same thing within braces too:

```
Boil the %pasta{1%cup|200%g} and…
```

I think commas are more useful for another purpose, specifying pre-prepared state:

```
add @green chilli{1, finely chopped}, @shallot{} or @red onion{1, finely chopped}, and @coriander leaves{1%tbsp, finely chopped}
```

**Reply by Zie0** [reply]
I think the pipe make sense though I might prefer two of them to make it a ["logical OR,"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR) but that might be overkill. IMO it helps to visually distinguish the two options at a glance though

For example, `(2 ½ cups / 370g)` might look something like `{2.5%cups||370%g}`.

And to approximate measurements, I think the [`~` to mean "about" is an easy way to get close to the math symbol for it](https://simple.wikipedia.org/wiki/List_of_mathematical_symbols). I often see `~` used for that on recipes, but I think that may be a separate discussion from the syntax for "OR"
